sdk/log/DESIGN.md

```diff
@@ -0,0 +1,170 @@
+# Logs SDK
+
+## Abstract
+
+`go.opentelemetry.io/otel/sdk/log` provides Logs SDK compliant with the
+[specification](https://opentelemetry.io/docs/specs/otel/logs/sdk/).
+
+The main and recommended use case is to configure the SDK to use an OTLP
+exporter with a batch processor.[^1] Therefore, the design aims to be
+high-performant in this scenario.
+
+The prototype was created in
+[#4955](https://github.com/open-telemetry/opentelemetry-go/pull/4955).
+
+## Modules structure
+
+The SDK is published as a single `go.opentelemetry.io/otel/sdk/log` Go module.
+
+The exporters are going to be published as following Go modules:
+
+- `go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploggrpc`
+- `go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp`
+- `go.opentelemetry.io/otel/exporters/stdout/stdoutlog`
+
+## LoggerProvider
+
+The [LoggerProvider](https://opentelemetry.io/docs/specs/otel/logs/sdk/#loggerprovider)
+is defined as follows:
+
+```go
+type LoggerProvider struct {
+	embedded.LoggerProvider
+}
+
+func NewLoggerProvider(...Option) *LoggerProvider
+
+func (*LoggerProvider) Logger(name string, options ...log.LoggerOption) log.Logger
+
+type Option interface { /* ... */ }
+
+func WithResource(*resource.Resource) Option
+```
+
+## LogRecord limits
+
+The [LogRecord limits](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecord-limits)
+are defined as follows:
+
+```go
+func WithAttributeCountLimit(limit int) Option
+
+func WithAttributeValueLengthLimit(limit int) Option
+```
+
+### LogRecordProcessor and LogRecordExporter  
+
+Both [LogRecordProcessor](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordprocessor)
+and [LogRecordExporter](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordexporter)
+are defined via an `Exporter` interface:
+
+```go
+func WithExporter(Exporter) Option {
+	return nil
+}
+
+type Exporter interface {
+	Export(ctx context.Context, records []*Record) error
```

Comment:
Thanks for the question. These are basically the things which I was thinking about during last week and I will do my best to describe my thought process.

> The SDK shouldn't ever call this with more than 1 `Record`, right?

By default this is correct, one log record would come in the beginning of the export pipeline. But one could use a `Batcher` (or its own custom implementation of a batching exporter) so that multiple records are in the pipeline.

> It seems like we need to split the concept of a processor versus an exporter given they will have different call patterns.
> Also, the processor should receive a mutable record, but the exporter should receive a batch of immutable records. This current design has both receiving a batch of mutable records.

The reasons why I propose to "merge" these two abstractions into a single `Exporter` interface are following.

This design gives more flexibility. E.g. it allows processing of the log records after they are batched.

I believe that having a representation of both mutable record and immutable record and converting from mutable to immutable may produce some performance overhead (probably minimal, but still - overhead). According to feedback in https://github.com/open-telemetry/opentelemetry-specification/pull/3898, I find it acceptable for an exporter to accept mutable log records. Notice that https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric#Exporter is also able to mutate the received metrics. 

The question is if it is acceptable for a processor to operate on multiple mutable records instead of a single record. However, take notice that if exporter will accept a mutable log records then (almost) everything what could be achieved by a custom processor can be achieved (and even more) by a custom exporter. I feel that having two interfaces would be redundant. Having both `Processor` and `Exporter` interfaces and structs for immutable and mutable log records would increase the API surface. Thus, I am proposing a simpler design with fewer abstraction where all functionalities required by the specification are available. For instance, if we would like to add `IsEnabled` support per https://github.com/open-telemetry/opentelemetry-specification/issues/3917 then we would need to only add this method to `Exporter` interface and not to both `Expoter` and `Processor` interfaces.

I admit that I also have doubts. This is why asked @tigrannajaryan, @jack-berg, and @open-telemetry/go-approvers   for feedback so early. However, I have a feeling that this a design worth exploring and that these simplifications are going to be profitable.

PS. I am putting this conversation as a reference to the design doc.

Reply:
CC @open-telemetry/specs-logs-approvers. I am open to your feedback as well.

Reply:
> The SDK shouldn't ever call this with more than 1 Record, right?

I have forgotten to mention that the exporter interface should say that slice must not be retained by the implementation (similarly to `io.Reader` and `io.Writer` implementations) so that the SDK can reuse a slice to avoid heap allocation. I will describe it tomorrow.

Reply:
I will look into introducing the Processor interface next week as the current way one would implement a processor by writing an exporter is not user friendly. e.g.

```go
var pool = sync.Pool{
	New: func() any {
		b := make([]Record, 0, 1)
		return &b
	},
}

type timestampDecorator struct {
	Exporter
}

func (e timestampDecorator) Export(ctx context.Context, records []Record) error {
	bPtr := pool.Get().(*[]Record)
	defer func() {
		*bPtr = (*bPtr)[:0]
		pool.Put(bPtr)
	}()
	b := *bPtr

	for _, r := range records {
		r = r.Clone()
		r.SetObservedTimestamp(time.Date(1988, time.November, 17, 0, 0, 0, 0, time.UTC))
		b = append(b, r)
	}

	return e.Exporter.Export(ctx, b)
}
```

Reply:
Addressed https://github.com/open-telemetry/opentelemetry-go/pull/4954/commits/f480907f5e067feab4b48b83f9b9465401f93187